### DISCUSSION TITLE
pls explain me this line I am beginner

### DISCUSSION THREAD

**Discussion by naviricbe** [community/community#35836] [Other Feature Feedback, Questions, & Ideas] [answered]
a = [1, -1, 3, -2]
for i in a:

    a[i] = a[i + - 1]          this line what it does and how
    
print(a)

result is [1,1,3,3] BUT HOW AND WHY

**Comment by JimNewaz** [comment]
i = index numbers of array 

`a[i + - 1]`  is similar to a[i - 1] 
Because we know, ( +  - ) = ( - )

So `a[i + - 1]` and `a[i - 1]` are same. 

Now, what does a[i-1] means? 
Here i and i-1 are the boundaries of the array and it will access two locations. 
i could be out of bounds, to begin with, or i-1 could be out of bounds resulting in an ArrayIndexOutOfBoundsException or a NullPointerException if you never initialized your array or i variable.

**Comment by Sachinsharma01** [comment]
Because 
 => + -  =  - 
 => - +  =  -
 => - - = +
 => + + = +

Note: the  resultant sign will be according to the bigger number

**Comment by DSDmark** [comment, marked as answer]
### First thing you need to understand in python, there are two ways to show or print, or you can say display the element of array.

![Screenshot from 2022-10-12 12-28-32](https://user-images.githubusercontent.com/85744816/195272101-ad1e0606-87b1-436a-9a02-28a029f7efac.png)

As you can see, index start with **0** and in the other hand you can also access with **-5**. You can access the elements of array with the **0** to **4** and the other hand **-5** to **-1**.

In for loop does not require an indexing variable to set beforehand.
But in your for loop program, we use ```i``` as an index to set the array elements. In other word, when you run that program of your

```
a = [1, -1, 3, -2]

for i in a:
    a[i] = a[i + - 1]    // the [i + - 1] is equal to a[ i - 1 ] as @JimNewaz describe
    
print(a)
```

in loop, the variable **i** contain all the elements of your array **a**.

for example:
```
for x in "banana":
  print(x) 

output:
b
a
n
a
n
a

```

When it's run the first time, it prints the value **1**.  Because the value of ```i``` is equal to the first element of the array, which is **1**  
so,
inside the loop is **a[i] = a[i+-1]**. The value of ```i``` is **1**, and it's look like this **a[1]=a[0]** and the **a[0]**  return **-1** because index of array start **0** so, first element of you array which is **1**.  It's actually like this **a[i + -1]** the value of ```i``` is **1**. First this **[i + -1]** solve which is ```1-1=0```. So, it's like that ```a[1] = a[0]```.
In second time when your loop is run the value of **i** is **-1** so, this expression look like this ```a[-1] = a[-2]```. The **a[-1}** as you can see in above images, **a[-1]** is the last index which is **-2** and **a[-2]** is the second last index which is **3** and it's look like this ```a[-1]=a[-2]```. and so on. All set the output you see like this :

```
[1, -1, 3, -2]      // previous array
[1, 1, 3, -2]       // changed array

[1, 1, 3, -2]       // previous array
[1, 1, 3, -2]      // changed array

[1, 1, 3, -2]       // previous array
[1, 1, 3, 3]         // change darray

[1, 1, 3, 3]        // previous array
[1, 1, 3, 3]         // changed array
```

**Reply by naviricbe** [reply]
Thanks !
